### PR TITLE
Add Testing section to CONTRIBUTING.md file

### PR DESCRIPTION
**Pull request by gkossakowski**
I realized that we don't have much information in `CONTRIBUTING.md` file about testing so I added some basic stuff.

Review by @retronym.

### CODE REVIEW

**Comment by xeno-by** [comment]
What is the relationship between CONTRIBUTING.md and http://scala-lang.org/contribute/hacker-guide.html#test? /cc @heathermiller

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -34,6 +34,30 @@ If *all* of these requirements are not met then the code should **not** be merge
 
 All contributed code should come accompanied with documentation. Pull requests containing undocumented code will not be accepted. Both user-facing Scaladoc comments, as well as committer-facing internal documentation (i.e. important design decisions that other maintainers should know about should be placed inline with line comments `//`) should be accompanying all contributed code where possible.
 
+## Testing
```
Comment (xeno-by) [inline CONTRIBUTING.md:37]: How does one turn on line wrapping in github?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -34,6 +34,30 @@ If *all* of these requirements are not met then the code should **not** be merge
...
+
+  * Scala library tests. E.g. all collection tests that assert certain behavior should be written as JUnit tests
+  * Symbol table tests that check correctness and contracts of Trees, Symbols and Types that do not require the whole compile to run (e.g. subtyping can be tested as an unit test)
+  * Single component of Scala compiler can and should be unit tested. For example, if you are writing an optimizer then it's a good idea to test each optimization in separation by passing some inputs (bytecode instructions) and asserting properties of outputs.
+
+In principle, every unit test can be written as partest's `run` test but there are several advantages of writing those tests as JUnit tests:
+
+  * JUnit has native IDE support so you can easily run and debug unit test in your favorite IDE
+  * JUnit tests use proper package structure (tests follow package structure of classes they test) so all tests can be compiled at once which dramatically speeds up compilation of tests
+  * JUnit reuses the same JVM instance to run all tests which speeds up tests execution
```
Comment (xeno-by) [inline CONTRIBUTING.md:58]: Is this safe wrt tests not affecting each other?

File: CONTRIBUTING.md (hunk 1 above)
Comment (retronym) [inline CONTRIBUTING.md:37]: We should reformat the sources to wrap, IMO.

**Comment by retronym** [comment]
partest has a nice ability to externalize assertions to a checkfile. This is really nice in some cases as it lets you see the results of many tests at once (e.g. [the results of a call at many compiler phases](https://github.com/retronym/scala/commit/1a98a5fee5603cae162d9359d97cf2f285753f7d#commitcomment-4039367)), and to update the expectations automatically when an expected change occurs. In many cases, the test becomes easier to understand, as you don't have to wade through a bunch of code making assertions about the structure of things, rather you can look at a textual representation of the expectation.

This is really orthogonal to the "Unit" vs "Functional" test distinction. I'd like us to think about whether we should provide the same facility to our JUnit tests.

We must also provide better command line support for running groups of tests from the command line.

**Comment by xeno-by** [comment]
@retronym +1

**Comment by gkossakowski** [comment]
Closing to reduce exposure to build bot issues. I'll resubmit whenever I have time.

**Pull request closed without merging**